events/handler.go

```diff
@@ -0,0 +1,124 @@
+package events
+
+import (
+	"fmt"
+)
+
+const (
+	// Event Types
+	LOSSY_FEED_EVENT = "LOSSY_FEED"
+
+	// Components
+	FEEDS_COMPONENT = "Feeds"
+)
+
+type Sink interface {
+	AddEvent(e Event) error
+}
+
+type Event interface {
+	GetComponent() string
+	GetType() string
+	GetMessage() string
+}
+
+type LossyFeedEvent struct {
+	Feed string
+
+	Event
+}
+
+type Filter struct {
+	enabledEventTypes  []string
+	disabledEventTypes []string
+
+	enabledComponents []string
+}
+
+type Handler struct {
+	eventSink   Sink
+	eventFilter Filter
+}
+
+func NewHandler(sink Sink, filter Filter) *Handler {
+	return &Handler{
+		eventSink:   sink,
+		eventFilter: filter,
+	}
+}
+
+func NewNullHandler() *Handler {
+	return &Handler{}
+}
+
+// Creates a filter for use with an event handler, nil can be provided for non values.
+func NewFilter(enabledEventTypes, disabledEventTypes, enabledComponents []string) *Filter {
+	if enabledEventTypes == nil {
+		enabledEventTypes = []string{}
+	}
+	if disabledEventTypes == nil {
+		disabledEventTypes = []string{}
+	}
+	if enabledComponents == nil {
+		enabledComponents = []string{}
+	}
+	return &Filter{
+		enabledEventTypes:  enabledEventTypes,
+		disabledEventTypes: disabledEventTypes,
+		enabledComponents:  enabledComponents,
+	}
+}
+
+// Dispatches an event to the configured sink if it passes the configured filter.
+// Filters are applied as follows:
+// 	- disabled event types are always disabled.
+//  - enabled event types are enabled
+//  - enabled components are enabled except for disabled event types
+func (h Handler) DispatchEvent(e Event) error {
+	if h.eventSink == nil {
+		return nil
+	}
+	dispatch := false
+	filter := h.eventFilter
+
+	eComponent := e.GetComponent()
+	eType := e.GetType()
+
+	// Enable components
+	if stringExistsInSlice(eComponent, filter.enabledComponents) {
+		dispatch = true
+	}
+	// Handle specific event types
+	if stringExistsInSlice(eType, filter.enabledEventTypes) {
+		dispatch = true
+	} else if stringExistsInSlice(eType, filter.disabledEventTypes) {
+		dispatch = false
+	}
+
+	if dispatch {
+		return h.eventSink.AddEvent(e)
+	}
+	return nil
+}
+
+// Checks for existance of a string within a slice of strings
+func stringExistsInSlice(s string, slice []string) bool {
+	for _, sliceStr := range slice {
+		if s == sliceStr {
+			return true
+		}
+	}
+	return false
+}
+
+func (e LossyFeedEvent) GetComponent() string {
+	return FEEDS_COMPONENT
+}
+
+func (e LossyFeedEvent) GetType() string {
+	return LOSSY_FEED_EVENT
+}
+
+func (e LossyFeedEvent) GetMessage() string {
```

Comment:
I wonder if for this specific event we'd want to expand the struct to store some of the date/time info which could be added to the log. This could be a useful datapoint for external investigation, for instance a given timeframe in which we can gather that no events were recorded? Obviously this could just be a false positive and an event may have genuinely not occurred during the delta , but it's at least a data point.

Reply:
I'm not sure whether we want this in the event structure itself but perhaps in the sink implementations, kafka for example comes with timestamping built in so we would be providing multiple timestamp values. Logrus could also be configured with a full timestamp field in the log message.

Reply:
I think that's subtly different to what I'm suggesting there. These would be separate timestamps pulled from the packages themselves (most recent package from previous results, oldest package from current results) as oppose to the timestamp at which the event itself was generated. This could be represented as a formatted string based field to avoid potential confusion

Reply:
As events are not formalised I think any extension to this or future event types can be proposed as followups, just to clarify that this is just a suggestion :smile: